javalib/src/main/scala/java/lang/Integer.scala

```diff
@@ -105,71 +157,420 @@ object Integer {
-    if (i == 0) 0 else if (i < 0) -1 else 1
-
-  // Intrinsic
-  def numberOfLeadingZeros(i: scala.Int): scala.Int = {
-    // See Hacker's Delight, Section 5-3
-    var x = i
-    if (x == 0) {
-      32
+    if (i == 0) 0
+    else if (i < 0) -1
+    else 1
+
+  @inline def sum(a: scala.Int, b: scala.Int): scala.Int =
+    a + b
+
+  def toBinaryString(i: scala.Int): String = {
+    // toUnsignedString(i, 1)
+    var count = 1
+    if (i < 0) {
+      count = 32
     } else {
-      var r = 1
-      if ((x & 0xffff0000) == 0) { x <<= 16; r += 16 }
-      if ((x & 0xff000000) == 0) { x <<= 8; r += 8 }
-      if ((x & 0xf0000000) == 0) { x <<= 4; r += 4 }
-      if ((x & 0xc0000000) == 0) { x <<= 2; r += 2 }
-      r + (x >> 31)
+      var j = i
+      while (j != 0) {
+        count += 1
+        j >>>= 1
+      }
     }
+
+    var k = i
+    val buffer = new Array[Char](count)
+    do {
+      count -= 1
+      buffer(count) = ((k & 1) + '0').toChar
+      k >>>= 1
+    } while (count > 0)
+
+    new String(buffer)
   }
 
-  @inline def numberOfTrailingZeros(i: scala.Int): scala.Int =
-    if (i == 0) 32
-    else 31 - numberOfLeadingZeros(i & -i)
+  def toHexString(i: scala.Int): String = {
+    // toUnsignedString(i, 4)
+    var count = 1
+    if (i < 0) {
+      count = 8
+    } else {
+      var j = i
+      while (j != 0) {
+        count += 1
+        j >>>= 4
+      }
+    }
+
+    var k = i
+    val buffer = new Array[Char](count)
+    do {
+      var t = k & 15
+      if (t > 9) {
+        t = t - 10 + 'a'
+      } else {
+        t += '0'
+      }
+      count -= 1
+      buffer(count) = t.toChar
+      k >>>= 4
+    } while (count > 0)
+
+    new String(buffer)
+  }
+
+  def toOctalString(i: scala.Int): String = {
+    // toUnsignedString(i, 3)
+    var count = 1
+    if (i < 0) {
+      count = 11
+    } else {
+      var j = i
+      while (j != 0) {
+        count += 1
+        j >>>= 3
+      }
+    }
+
+    var k = i
+    val buffer = new Array[Char](count)
+    do {
+      count -= 1
+      buffer(count) = ((k & 7) + '0').toChar
+      k >>>= 3
+    } while (count > 0)
+
+    new String(buffer)
+  }
+
+  @inline def toString(i: scala.Int): String = {
+    if (i == 0) {
+      "0"
+    } else {
+      val negative = i < 0
+
+      if (i < 1000 && i > -1000) {
+        val buffer = new Array[Char](4)
+        val positive_value =
+          if (negative) -i
+          else i
+        var first_digit = 0
+        if (negative) {
+          buffer(0) = '-'
+          first_digit += 1
+        }
+
+        var last_digit = first_digit
+        var quot       = positive_value
+        do {
+          val res = quot / 10
+          var digit_value = quot - ((res << 3) + (res << 1))
+          digit_value += '0'
+          last_digit += 1
+          buffer(last_digit) = digit_value.toChar
+          quot = res
+        } while (quot != 0)
+
+        last_digit -= 1
+        val count = last_digit
```

Comment:
Here is the same mistake, `last_digit--` means that you first read previous value and then decrement. `val count = last_digit` needs to come before `last_digit -= 1`.